cpp/src/neighbors/detail/cagra/search_single_cta_kernel-inl.cuh

```diff
@@ -905,6 +1191,728 @@ struct search_kernel_config {
+      job_queue.capacity(),
+      worker_queue.capacity());
+    last_touch.store(std::chrono::system_clock::now(), std::memory_order_relaxed);
+  }
+
+  ~persistent_runner_t() noexcept override
+  {
+    auto whs = worker_handles.data();
+    for (auto i = worker_handles.size(); i > 0; i--) {
+      whs[worker_queue.pop().wait()].data.store({kNoMoreWork}, cuda::memory_order_relaxed);
+    }
+    RAFT_CUDA_TRY_NO_THROW(cudaStreamSynchronize(stream));
+    RAFT_LOG_INFO("Destroyed the persistent runner.");
+  }
+
+  void launch(index_type* result_indices_ptr,       // [num_queries, top_k]
+              distance_type* result_distances_ptr,  // [num_queries, top_k]
+              const data_type* queries_ptr,         // [num_queries, dataset_dim]
+              uint32_t num_queries,
+              uint32_t top_k)
+  {
+    // submit all queries
+    launcher_t launcher{
+      job_queue, worker_queue, worker_handles.data(), num_queries, [=](uint32_t job_ix) {
+        auto& jd                = job_descriptors.data()[job_ix].input.value;
+        auto* cflag             = &job_descriptors.data()[job_ix].completion_flag;
+        jd.result_indices_ptr   = result_indices_ptr;
+        jd.result_distances_ptr = result_distances_ptr;
+        jd.queries_ptr          = queries_ptr;
+        jd.top_k                = top_k;
+        jd.n_queries            = num_queries;
+        cflag->store(false, cuda::memory_order_relaxed);
+        cuda::atomic_thread_fence(cuda::memory_order_release, cuda::thread_scope_system);
+        return cflag;
+      }};
+
+    // Update the state of the keep-alive atomic in the meanwhile
+    auto prev_touch = last_touch.load(std::memory_order_relaxed);
+    if (prev_touch + kLiveInterval / 10 < launcher.now) {
+      // to avoid congestion at this atomic, we only update it if a significant fraction of the live
+      // interval has passed.
+      last_touch.store(launcher.now, std::memory_order_relaxed);
+    }
+    // wait for the results to arrive
+    launcher.wait();
+  }
+
+  auto calc_coop_grid_size(uint32_t block_size, uint32_t smem_size) -> dim3
+  {
+    // We may need to run other kernels alongside this persistent kernel.
+    // So we can leave a few SMs idle.
+    // Note: running any other work on GPU alongside with the persistent kernel make the setup
+    // fragile.
+    //   - Running another kernel in another thread usually works, but no progress guaranteed
+    //   - Any CUDA allocations block the context (this issue may be obscured by using pools)
+    //   - Memory copies to not-pinned host memory may block the context
+    //
+    // Even when we know there are no other kernels working at the same time, setting
+    // kDeviceUsage to 1.0 surprisingly sometimes hurts performance. Proceed with care.
+    // If you suspect this is an issue, you can reduce this number to ~0.9 without a significant
+    // impact on the throughput.
+    constexpr double kDeviceUsage = 1.0;
```

Comment:
The comments above suggest, that it could make sense to make this a runtime parameter. (Although I do not insist to do so if it is a very rare case that running less than full device is better). Alternatively, could we use `thread_block_size`? already defined in `search_params` struct? IIUC thread_block_size is ignored now, and we derive number of blocks using this kDeviceUsage and the device parameters that we query below. Is that correct?

Reply:
Yes, it makes sense to make it a runtime parameter. I'd like to wait till https://github.com/rapidsai/cuvs/pull/227 is merged though, because we may end up having quite a few parameters :)
As for the other question: `thread_block_size` is not directly related: I do use it to find out the occupancy (what's the maximum grid size); then `kDeviceUsage` takes a fraction of the found maximum grid size.